### PR TITLE
feat(launch): add --timeout / --retries flags with idle-stdout watchdog

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Adds two new flags to `ezpz launch` to mitigate silent training hangs.

```bash
# Kill if stdout goes quiet for 10 minutes; retry up to 3 times.
ezpz launch --timeout 600 --retries 3 -- python3 -m my_app.train
```

## Motivation

`torchtitan`'s `train_timeout_seconds` is silently ignored on Intel
XPU because PyTorch's `_set_pg_timeout` has no `xpu` branch — every
xccl process group falls through to a no-op warning at startup
("`Set timeout is now only supported for either nccl or gloo.`")
and the configured per-collective timeout never takes effect.

The practical consequence: an XPU job that hangs in a collective
(EP all-to-all desync, fabric flap, single-rank stall) sits silent
until PBS walltime expires — burning ~12h of node-hours instead of
aborting at the configured ~100s. The warning is the only signal,
and it appears in every run, so it gets tuned out.

The launch-side watchdog is the *process-level* defense — it doesn't
depend on PyTorch's dispatch and works for any backend. Two
companion follow-ups (config-time warning, in-process step-progress
watchdog) are noted in the source code writeup but out of scope here.

## New flags

| Flag | Default | Semantics |
|---|---|---|
| `--timeout SECONDS` | off | Send SIGTERM → 10s grace → SIGKILL if no stdout for SECONDS. Exit 124 (GNU `timeout(1)` convention). *Idle, not walltime.* |
| `--retries N` | 0 | Re-execute on any non-zero exit, up to N times. Exponential backoff: 5s, 10s, 20s, 40s, capped 60s. |

## Implementation

- `_run_with_watchdog(cmd, idle_timeout_s)` in `src/ezpz/launch.py`.
  Fast-paths to `subprocess.run` when timeout is `None`. Otherwise
  Popen with merged stdout/stderr, background reader thread pumps
  to our stdout and updates a monotonic activity timestamp; main
  thread polls for process exit or idle deadline.
- `_run_with_retries(cmd, idle_timeout_s, retries)` wraps it in a
  retry loop with clear per-attempt logging.
- Wired through both call sites in `launch.py`: the
  scheduler-detected `launch()` path AND the local-mpirun `run()`
  fallback. Also fixes a latent bug where `launch()`'s return code
  was discarded in the scheduler branch.

## Tests

`tests/test_launch_watchdog.py` — 11 tests covering:

- Passthrough modes (`timeout=None`, `timeout=0`)
- Silent process is killed within timeout + grace
- Chatty process (echoes every 0.2s under a 1s timeout) is NOT killed
- Exit code propagation (clean exit 0, non-zero exit 42)
- Retry loop: zero retries = single attempt, success first try,
  fail-then-succeed (using a counter file), retries exhausted
- Watchdog + retry composition (watchdog kill is a retryable failure)

All 11 pass locally in <10 seconds.

## Scope caveat (documented in source + docs)

The watchdog only watches the process `ezpz launch` spawns directly.
If a user `qsub`s a job script that internally invokes
`python train.py`, the watchdog needs to live inside *that* script,
not the outer `qsub`. The in-process step-progress watchdog
(separate follow-up) is the complementary defense for that case.

## Test plan

- [x] `pytest tests/test_launch_watchdog.py` — 11/11 pass
- [x] `pytest tests/test_kill.py tests/test_yeet_env.py` — no regressions
- [x] `ezpz launch --help` shows both flags with full help text
- [x] `parse_args(['--timeout', '300', '--retries', '2', '--', 'python', 'train.py'])` round-trips correctly
- [ ] On a real ALCF job: `ezpz launch --timeout 60 -- sleep 120` aborts at 60s with exit 124
- [ ] Same with `--retries 1`: two SIGTERM cycles, then exits 124

## Summary by Sourcery

Add configurable idle-stdout watchdog and retry support to `ezpz launch`, wiring them through the CLI, launcher paths, and documentation, with focused tests covering the new behavior.

New Features:
- Introduce an idle-stdout watchdog for launched processes, configurable via a new `--timeout` flag and returning exit code 124 when it triggers.
- Add a retry mechanism for failed launches, configurable via a new `--retries` flag with exponential backoff between attempts.

Enhancements:
- Ensure scheduler-based and local-mpirun launch paths both honor watchdog and retry settings and correctly propagate child exit codes.

Documentation:
- Document the new `--timeout` and `--retries` flags, including semantics, caveats, and usage examples in the `ezpz launch` CLI docs.

Tests:
- Add tests for the idle-stdout watchdog and retry loop behaviors, including passthrough modes, hang detection, exit code propagation, and watchdog–retry interaction.

### CODE REVIEW

File: src/ezpz/launch.py [hunk 1]
```diff
@@ -24,6 +25,151 @@
...
+    if idle_timeout_s is None or idle_timeout_s <= 0:
+        proc = subprocess.run(cmd, check=False)
+        return proc.returncode
+
+    # Merge stderr into stdout so a single reader thread covers both.
+    process = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py:87]: Fixed in 938a306. Set `PYTHONUNBUFFERED=1` in the child env via `Popen(env=...)` (with `setdefault` so user-set values aren't clobbered). Two regression tests added: `test_python_child_is_unbuffered` and `test_python_unbuffered_can_be_overridden`. The whole reason this was P1 — perfectly healthy Python jobs being SIGTERMed because their 4-8KB stdout buffer hadn't flushed — is now covered.

File: src/ezpz/launch.py [hunk 2]
```diff
@@ -24,6 +25,151 @@
...
+# Grace period between SIGTERM and SIGKILL when the watchdog fires.
+# Gives well-behaved processes a chance to flush logs and shut down
+# cleanly; SIGKILL is the hammer for anything stuck.
+_WATCHDOG_KILL_GRACE_S = 10.0
+
+# Backoff cap between retry attempts (5s, 10s, 20s, 40s, then capped).
+_RETRY_BACKOFF_CAP_S = 60.0
+
+
+def _run_with_watchdog(
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py:42]: Partially addressed in 938a306. Added the suggested `try/finally` around the polling loop so `reader_done.wait(timeout=2.0)` runs uniformly on all exit paths (normal exit, watchdog kill, exception). Skipped the `_log_and_sleep_before_retry` extraction — `_run_with_retries` is ~15 lines with one branch, and pulling out the logging/backoff feels like overkill at that size. Happy to revisit if it grows.

File: tests/test_launch_watchdog.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    def test_retry_works_with_watchdog_kill(self, monkeypatch):
+        """The watchdog's 124 exit code is a non-zero exit like any
+        other, so retries should kick in. Verifies the two helpers
+        compose cleanly."""
+        # First attempt: sleep forever (gets killed by watchdog → 124).
+        # We can't easily make subsequent attempts behave differently
+        # without a state file, so just verify that watchdog-killed
+        # processes are eligible for retry: set retries=1 and observe
+        # that we attempted twice (each watchdog kill takes ~1s + grace).
+        monkeypatch.setattr("ezpz.launch.time.sleep", lambda *_a, **_k: None)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_launch_watchdog.py]: Addressed in b0aa6a5. The test now uses a counter file (like the other retry tests) and directly asserts `counter.read_text().strip() == "2"` for `retries=1`. No more timing-based inference.

File: src/ezpz/launch.py [hunk 4]
```diff
@@ -24,6 +25,151 @@
...
+        return proc.returncode
+
+    # Merge stderr into stdout so a single reader thread covers both.
+    process = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,  # line-buffered
+    )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py:90]: Acknowledged but not actionable — this is the standard subprocess-with-list-arg pattern (`cmd` is a `Sequence[str]` from `shlex.split`/argparse, not a shell string), and `shell=False` is the default. `shlex.escape` doesn't apply here since we're not constructing a shell command line; we're passing argv directly to `execve`. The command source is the user's own CLI invocation, so there's no untrusted external input to escape. False positive from the opengrep rule.

File: src/ezpz/cli/flags.py [hunk 5]
```diff
@@ -341,6 +341,33 @@ def build_launch_parser(
...
+        "--retries",
+        type=int,
+        default=0,
+        dest="retries",
+        help=(
+            "Re-execute the launched command up to N times on non-zero "
+            "exit. Applies exponential backoff between attempts "
+            "(5s, 10s, 20s, ..., capped at 60s). Default: 0 (no retry)."
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/cli/flags.py:393]: Fixed in b0aa6a5. Added a `_non_negative_int` argparse type validator that raises `argument --timeout: must be >= 0 (got -5)` at parse time. Applied to both `--timeout` and `--retries`. `0` is still accepted (and explicitly documented as "disabled" / "no retry").

File: src/ezpz/cli/flags.py [hunk 6]
```diff
@@ -341,6 +341,33 @@ def build_launch_parser(
...
+        dest="idle_timeout_s",
+        help=(
+            "Idle-stdout watchdog timeout in seconds. If the launched "
+            "process produces no stdout for this many seconds, send "
+            "SIGTERM (then SIGKILL after a 10s grace period) and exit "
+            "with code 124. NOT a total walltime — the process can run "
+            "indefinitely as long as it keeps emitting output. Off by "
+            "default. Useful for catching collective hangs (e.g. xccl "
+            "silent deadlock on XPU) that would otherwise consume the "
+            "full PBS walltime."
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/cli/flags.py]: Fixed in b0aa6a5. Help text now reads "no output (on stdout OR stderr — they are merged)".

File: src/ezpz/launch.py [hunk 7]
```diff
@@ -24,6 +25,151 @@
...
+    process's stdout in real time, and SIGTERM it (then SIGKILL after
+    ``_WATCHDOG_KILL_GRACE_S``) if no output appears for
+    ``idle_timeout_s`` consecutive seconds. Returns the process's
+    exit code, or ``_WATCHDOG_EXIT_CODE`` (124) if the watchdog fired.
+
+    Idle = stdout silence. The process can run indefinitely as long
+    as it keeps emitting at least one line per ``idle_timeout_s``.
+    Designed for catching collective hangs (xccl, NCCL, etc.) where
+    the process is alive but every rank is blocked in the same
+    collective and nothing reaches stdout.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py]: Fixed in b0aa6a5. Updated the docstring to say "Idle = silence on *both* stdout and stderr (stderr is merged into stdout)" so the wording matches the implementation.

File: tests/test_launch_watchdog.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+  * ``--timeout SECONDS`` — kill the launched process if stdout goes
+    silent for this many seconds (motivating case: xccl on XPU
+    silently ignores ``train_timeout_seconds`` and PBS jobs sit hung
+    for the full walltime).
+
+  * ``--retries N`` — re-execute on any non-zero exit, up to N times.
+
+All tests use ``sh -c`` scripts to simulate child behavior so they
+run on any platform without depending on Python subprocess timing.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_launch_watchdog.py]: Fixed in b0aa6a5. Added module-level `pytestmark = pytest.mark.skipif(os.name != "posix", ...)` so the whole file skips cleanly on Windows. Also updated the docstring to acknowledge POSIX-only — ezpz launch targets PBS/SLURM/mpirun environments which are all POSIX, so a full cross-platform rewrite isn't justified for these tests.

File: docs/cli/launch/index.md [hunk 9]
```diff
@@ -82,8 +82,56 @@ allocated to your job.
...
 
+## Idle-stdout watchdog (`--timeout`)
+
+`--timeout SECONDS` arms a watchdog that monitors the launched
+process's stdout. If no output appears for `SECONDS` consecutive
+seconds, the watchdog sends `SIGTERM`, waits up to 10 seconds for a
+clean shutdown, then sends `SIGKILL`. The exit code returned by
+`ezpz launch` is `124` (matching GNU `timeout(1)` convention) so
+shell wrappers can distinguish "killed for going silent" from
+"command failed".
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/cli/launch/index.md]: Fixed in b0aa6a5. Updated the section to "Idle-output watchdog", clarified that the watchdog monitors **stdout or stderr** (they're merged), and added a new "Python buffering" subsection explaining the `PYTHONUNBUFFERED=1` behavior so users understand what the watchdog touches.
